### PR TITLE
feat: sync-to-latest-tests

### PR DESCRIPTION
**Pull request by TheAustrian1998**
- add tests for sync function
- fix errors

### CODE REVIEW

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 1]
```diff
@@ -445,4 +456,211 @@ contract('DCAStrategiesPositionsHandler', () => {
...
+
+          newPositions = (await DCAStrategiesPositionsHandlerMock.userPosition(1)).positions;
+        });
+        then('transfer() is called correctly', async () => {
+          expect(tokenF.transfer).to.have.been.calledWith(user.address, delta);
+        });
+        then('increase, reduce, deposit or terminate is called correctly', async () => {
+          checkCallsToHub();
+          expect(hub.reducePosition).to.have.been.calledTwice;
+          expect(hub.increasePosition).to.have.callCount(0);
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:544]:
```suggestion
          expect(hub.increasePosition).to.not.have.been.called
```

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 2]
```diff
@@ -445,4 +456,211 @@ contract('DCAStrategiesPositionsHandler', () => {
...
+          tx = await callSync(totalAmount.sub(delta), newAmountOfSwaps);
+
+          newPositions = (await DCAStrategiesPositionsHandlerMock.userPosition(1)).positions;
+        });
+        then('transfer() is called correctly', async () => {
+          expect(tokenF.transfer).to.have.been.calledWith(user.address, delta);
+        });
+        then('increase, reduce, deposit or terminate is called correctly', async () => {
+          checkCallsToHub();
+          expect(hub.reducePosition).to.have.been.calledTwice;
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:543]: Could we check that each reduce was called with the correct amount? I know it's a pain, but I think it's very important to understand that we are doing everything correctly

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 3]
```diff
@@ -445,4 +456,211 @@ contract('DCAStrategiesPositionsHandler', () => {
...
+        given(async () => {
+          tx = await callSync(totalAmount.sub(delta), newAmountOfSwaps);
+
+          newPositions = (await DCAStrategiesPositionsHandlerMock.userPosition(1)).positions;
+        });
+        then('transfer() is called correctly', async () => {
+          expect(tokenF.transfer).to.have.been.calledWith(user.address, delta);
+        });
+        then('increase, reduce, deposit or terminate is called correctly', async () => {
+          checkCallsToHub();
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:542]: Could we check that `deposit` and `terminate` were called with the correct parameters?

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 4]
```diff
@@ -445,4 +456,211 @@ contract('DCAStrategiesPositionsHandler', () => {
...
+        ).to.be.revertedWith('NoPermissions()');
+      });
+    });
+    when('syncPositionToNewVersion is called', () => {
+      given(async () => {
+        const positions = [1, 2, 3];
+        const OLD_SHARE_A = { token: tokenA.address, share: BigNumber.from(33e2) }; // reduce
+        const OLD_SHARE_B = { token: tokenB.address, share: BigNumber.from(33e2) }; // increase
+        const OLD_SHARE_C = { token: tokenC.address, share: BigNumber.from(34e2) }; // terminate
+        oldShares = sortTokens([OLD_SHARE_A, OLD_SHARE_B, OLD_SHARE_C]);
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:486]: Could we add a token whose shares doesn't change? It would be interesting to understand that it works correctly

**Review by nchamo — COMMENTED** [review]
Amazing work @TheAustrian1998 !

Left some comments

File: test/unit/DCAStrategies/dca-strategies-management-handler.spec.ts [hunk 5]
```diff
@@ -368,12 +368,19 @@ contract('DCAStrategiesManagementHandler', () => {
...
-      if (parseInt(a.token, 16) > parseInt(b.token, 16)) return 1;
+      if (hexToNumber(a.token) < hexToNumber(b.token)) return -1;
+      if (hexToNumber(a.token) > hexToNumber(b.token)) return 1;
+      if (hexToNumber(a.token) == hexToNumber(b.token)) console.error('found duplicate when sorting');
       return 0;
     }
 
-    return array.sort(compare);
+    let f = array.sort(compare);
+    console.log('array f', f);
```
[the remark this answers is not in the document]
Comment (TheAustrian1998) [reply test/unit/DCAStrategies/dca-strategies-management-handler.spec.ts:383]: My bad

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts (hunk 4 above)
Comment (TheAustrian1998) [reply test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:486]: good catch, I wasn't accounting for a edge-case where nothing changes

File: contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol [hunk 6]
```diff
@@ -371,6 +372,9 @@ abstract contract DCAStrategiesPositionsHandler is IDCAStrategiesPositionsHandle
             positionId: _currentPositionId,
             amount: _correspondingToPosition - _userPosition.remaining
           });
+        } else {
+          // do nothing
+          _tasks[_data.newPositionsIndex] = Task({action: Action.NOTHING, positionId: _currentPositionId, amount: 0});
```
Comment (nchamo) [inline contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:377]: Is this necessary? Since `NOTHING` is the first value in the enum, it will be the default value, so I think it might not be necessary to assign it

File: test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts [hunk 7]
```diff
@@ -445,4 +458,256 @@ contract('DCAStrategiesPositionsHandler', () => {
...
+      then('tx reverts with message', async () => {
+        await expect(
+          DCAStrategiesPositionsHandlerMock.syncPositionToNewVersion(1, 3, user.address, user.address, amountOfSwaps, newAmountOfSwaps)
+        ).to.be.revertedWith('NoPermissions()');
+      });
+    });
+    when('syncPositionToNewVersion is called', () => {
+      given(async () => {
+        const positions = [1, 2, 3];
+        const OLD_SHARE_A = { token: tokenA.address, share: BigNumber.from(33e2) }; // nothing
```
Comment (nchamo) [inline test/unit/DCAStrategies/dca-strategies-positions-handler.spec.ts:486]: But now we are not testing reduce 😅 

Could we try to test all situations? Maybe adding an extra token?

**Review by nchamo — COMMENTED** [review]
Minor comments

File: contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol (hunk 6 above)
Comment (TheAustrian1998) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:377]: I think it is. I need it to save the positionId to be assigned in the last for loop. I can reuse existing enum values if you want. (In line 377 when writing the task I could use the REDUCE value or any other. It doesn't matter because is not read).

File: contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol [hunk 8]
```diff
@@ -241,19 +247,17 @@ abstract contract DCAStrategiesPositionsHandler is IDCAStrategiesPositionsHandle
...
       }
 
+      _newPositions[i] = _task.positionId;
+
       unchecked {
         i++;
       }
     }
 
+    _userPositions[_positionId].positions = _newPositions;
```
[the remark this answers is not in the document]
Comment (TheAustrian1998) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:259]: I noticed some strange behaviors that caused an ID to be overwritten (i.e. a position that was reused was overridden by a new position just deposited). Also, it made the code much more simpler 😅

[the remark this answers is not in the document]
Comment (nchamo) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:259]: I agree that the code is **much much** simpler, but I think it will end up being more expensive. Because in the case of increase/reduce/nothing, you are re-writing the same value that was already there

[the remark this answers is not in the document]
Comment (TheAustrian1998) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:259]: What do you suggest?

File: contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol [hunk 9]
```diff
@@ -229,7 +229,7 @@ abstract contract DCAStrategiesPositionsHandler is IDCAStrategiesPositionsHandle
       IERC20(_data.positionMetadata.from).safeTransfer(_recipientUnswapped, _data.totalRemaining - _totalAmount);
     }
 
-    uint256[] memory _newPositions = new uint256[](_tasks.length);
+    uint256 _auxPositionId = _positionId;
```
[the remark this answers is not in the document]
Comment (TheAustrian1998) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:232]: Remember that weird `Stack too deep` error? That's the way of fixing it

[the remark this answers is not in the document]
Comment (nchamo) [reply contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol:232]: OMG Solidity, why you like this?

**Review by nchamo — APPROVED** [review]
Amazing work @TheAustrian1998 !

This was a really long and hard PR, but you make it work 💪💪💪 

Thank you so much for this!
